File: engines/nodejs14/benchmark.js
#!/usr/bin/env node

/**
 * Node.js 14 Benchmark Program
 * Uses traditional Node.js features for compatibility
 */

const fs = require('fs');
const { performance } = require('perf_hooks');

function main() {
    // Check command line arguments
    if (process.argv.length !== 5) {
        console.error(`Usage: node ${process.argv[1]} <base64_regex> <filename> <match_mode>`);
        console.error('  base64_regex: Base64-encoded regular expression');
        console.error('  filename: Path to the file containing text to match');
        console.error('  match_mode: 1 for full match, 0 for partial match');
        process.exit(1);
    }

    const base64Regex = process.argv[2];
    const filename = process.argv[3];
    const matchModeStr = process.argv[4];

    try {
        // Decode the base64 regex
        const regex = decodeBase64(base64Regex);
        
        // Parse match mode
        const matchMode = parseMatchMode(matchModeStr);
        
        // Read file content
        const data = readFileContent(filename);
        
        // Measure and output results
        measurePerformance(data, regex, matchMode === 1);
        
    } catch (error) {
        console.error('Error:', error.message);
        process.exit(1);
    }
}

function decodeBase64(base64String) {
    try {
        if (!base64String || base64String.length === 0) {
            throw new Error('Base64 string cannot be empty');
        }
        
        const decoded = Buffer.from(base64String, 'base64').toString('utf8');
        
        if (decoded.length === 0) {
            throw new Error('Decoded regex cannot be empty');
        }
        
        return decoded;
    } catch (error) {
        throw new Error(`Failed to decode base64 regex: ${error.message}`);
    }
}

function readFileContent(filename) {
    try {
        return fs.readFileSync(filename, { encoding: 'utf8' });
    } catch (error) {
        throw new Error(`Cannot open file ${filename}: ${error.message}`);
    }
}

function parseMatchMode(matchModeStr) {
    const matchMode = parseInt(matchModeStr, 10);
    
    if (isNaN(matchMode) || (matchMode !== 0 && matchMode !== 1)) {
        throw new Error('match_mode must be 0 or 1');
    }
    
    return matchMode;
}

function measurePerformance(data, patternStr, fullMatch) {
    const startTime = performance.now();

    try {
        // Create regex with appropriate flags
        const flags = fullMatch ? '' : 'g';
        const pattern = new RegExp(patternStr, flags);
        let count = 0;

        if (fullMatch) {
            // Full match: entire text must match the regex (use ^ and $ anchors)
            const fullPattern = new RegExp('^(?:' + patternStr + ')$');
            count = fullPattern.test(data.trim()) ? 1 : 0;
        } else {
            // Partial match: find all matches using compatible approach
            const matches = data.match(new RegExp(patternStr, 'g'));
            count = matches ? matches.length : 0;
        }

        const endTime = performance.now();
        const elapsedMs = endTime - startTime;

        // Output with high precision formatting
        console.log(`${elapsedMs.toFixed(6)} - ${count}`);

    } catch (error) {
        throw new Error(`Failed to compile regex: ${error.message}`);
    }
}

// Run the main function if this module is executed directly
if (require.main === module) {
    main();
}

File: engines/java11/Benchmark.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.time.Instant;
import java.util.Base64;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class Benchmark {
    public static void main(String... args) throws IOException {
        if (args.length != 3) {
            System.out.printf("Usage: java %s <base64_regex> <filename> <match_mode>%n", 
                             Benchmark.class.getSimpleName());
            System.out.println("  base64_regex: Base64-encoded regular expression");
            System.out.println("  filename: Path to the file containing text to match");
            System.out.println("  match_mode: 1 for full match, 0 for partial match");
            System.exit(1);
        }

        try {
            // Decode the base64 regex using modern Java
            var regex = decodeBase64(args[0]);
            
            // Read file content using Java 11 enhanced file I/O
            var data = readFile(args[1]);
            
            // Parse match mode
            var matchMode = parseMatchMode(args[2]);
            
            // Measure and output results
            measure(data, regex, matchMode == 1);
            
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            System.exit(1);
        }
    }

    private static String decodeBase64(String base64String) {
        try {
            var decoded = Base64.getDecoder().decode(base64String);
            return new String(decoded);
        } catch (IllegalArgumentException e) {
            throw new RuntimeException("Failed to decode base64 regex: " + e.getMessage());
        }
    }

    private static String readFile(String filename) throws IOException {
        try {
            var path = Path.of(filename);
            return Files.readString(path);
        } catch (IOException e) {
            throw new IOException("Cannot open file " + filename + ": " + e.getMessage());
        }
    }

    private static int parseMatchMode(String matchModeStr) {
        try {
            var matchMode = Integer.parseInt(matchModeStr);
            if (matchMode != 0 && matchMode != 1) {
                throw new IllegalArgumentException("match_mode must be 0 or 1");
            }
            return matchMode;
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException("match_mode must be 0 or 1");
        }
    }

    private static void measure(String data, String patternStr, boolean fullMatch) {
        var startTime = Instant.now();

        try {
            var pattern = Pattern.compile(patternStr);
            var count = 0;

            if (fullMatch) {
                // Full match: entire text must match the regex
                var matcher = pattern.matcher(data);
                if (matcher.matches()) {
                    count = 1;
                } else {
                    count = 0;
                }
            } else {
                // Partial match: find all matches in the text
                var matcher = pattern.matcher(data);
                while (matcher.find()) {
                    count++;
                }
            }

            var endTime = Instant.now();
            var duration = Duration.between(startTime, endTime);
            var elapsedMs = duration.toNanos() / 1_000_000.0;

            System.out.printf("%.6f - %d%n", elapsedMs, count);

        } catch (Exception e) {
            System.err.println("Error: Failed to compile regex: " + e.getMessage());
            System.exit(1);
        }
    }
} 